### DISCUSSION TITLE
GPUCompilationMessage for "cyclic dependency" is weird

### DISCUSSION THREAD

**Discussion by stefnotch** [gpuweb/gpuweb#4611] [Q&A] [answered]
A GPUCompilationMessage has a `linePos` that points at the line which caused the error.

However, if we have a cyclic dependency, like the following

```wgsl
fn foo() -> u32 {
   return bar();
}
fn bar()  -> u32 {
   return baz();
}
fn baz() -> u32 {
return foo();
}
```

then which line caused the error? What should the GPUCompilationMessage's linePos say?

Both [naga](https://github.com/gfx-rs/wgpu/blob/08efa72a83d0acdf25177408150e4c87933c1cd7/naga/src/front/wgsl/index.rs#L148-L159) and [Chrome's Tint](https://dawn.googlesource.com/tint/+/refs/heads/main/docs/tint/diagnostics_guide.md) can internally generate multiple locations for this and a few other errors.

For what it's worth, the text of the error message will indicate where the error happened. However, websites like Shadertoy and https://compute.toys/ would have to
- carefully parse the returned error message (is that even legal?)
- or include naga and get the error messages from there
```
Compilation log for [Invalid ShaderModule (unlabeled)]:
1 error(s) generated while compiling the shader:
:79:1 error: cyclic dependency found: 'foo' -> 'bar' -> 'baz' -> 'foo'
fn foo() -> u32 {
^^^^^^^^^^^^^^^^^
   return bar();
^^^^^^^^^^^^^^^^
}
^

:80:11 note: function 'foo' references function 'bar' here
   return bar();
          ^^^

:83:11 note: function 'bar' references function 'baz' here
   return baz();
          ^^^

:86:8 note: function 'baz' references function 'foo' here
return foo();
       ^^^
```

**Comment by greggman** [comment, marked as answer]
I'm not sure what you're asking for. AFAICT, the spec requires the WGSL to fail but it doesn't require any specific error message or line numbers. Those are entirely undefined and up to the implementation. It's fine for it to just say "bad WGSL" and nothing else. That's intentional as it is in every other GPU API since they may all take different paths to get to an error. All that's required is they don't allow invalid code.

That said, I'm not sure what you're asking for that's not already there. When I run this in Chrome I  I call [`module.getCompilationInfo`](https://www.w3.org/TR/webgpu/#dom-gpushadermodule-getcompilationinfo) and it returns all the data required to highlight each line, no parsing required.

<img width="467" alt="Screenshot 2024-04-29 at 23 11 27" src="https://github.com/gpuweb/gpuweb/assets/234804/98f8f57b-f51d-4550-ab2c-4d82403ce85f">

And here's an example highlighting each line based on that info

[example](https://webgpufundamentals.org/webgpu/lessons/resources/wgsl-offset-computer.html#x=5d00000100b200000000000000003d888b0237284ce7dd1fddbe8c660e7f3da1e3fe4221a7ab9badfd76c48aaf8abe0fa2c1258e229e3a0ed058bca38d581fa9415c8e554ae76408f38c74f469388bb84913c8411ffe9625b8568cb8f78f05cf764efc25ac2241caac4ea9462b1e3a2b382ccc6479e1c0a3cc6ed630758abe2949deb8c7989c238cd4bffeb1fdc0)

**Reply by stefnotch** [reply]
I was indirectly trying to ask "does the GPUCompilationMessage struct actually let all browser vendors include enough info, or do they end up throwing away potentially quite useful info".

Thank you for the Chrome example, I suppose returning multiple messages is a quite reasonable option.

**Comment by arjami25** [comment]
thanks for that..
